libr/bin/p/bin_dex.c

```diff
@@ -1567,85 +1574,42 @@ static void parse_class(RBinFile *bf, RBinDexClass *c, int class_index, int *met
 		    c->class_data_offset <
 			    dex->header.class_offset +
 				    dex->header.class_size * DEX_CLASS_SIZE) {
-			return;
+			goto beach;
 		}
 
-		ut64 bufbufsz;
-		const ut8 *bufbuf = r_buf_data (bf->buf, &bufbufsz);
-		p = bufbuf + c->class_data_offset;
-		// XXX may overflow
-		if (bufbufsz < c->class_data_offset) {
-			return;
+		RBinDexClassData *dc = R_NEW0 (RBinDexClassData);
+		if (!dc) {
+			goto beach;
 		}
-		ut32 p_size = (bufbufsz - c->class_data_offset);
-		p_end = p + p_size;
-		//XXX check for NULL!!
-		c->class_data = (struct dex_class_data_item_t *)malloc (
-			sizeof (struct dex_class_data_item_t));
-		if (!c->class_data) {
-			return;
-		}
-		if (p >= p_end) {
-			free (c->class_data);
-			return;
-		}
-		ut64 eof;
 
-		p = r_uleb128 (p, p_end - p, &eof);
-		if (p >= p_end) {
-			free (c->class_data);
-			return;
+		bool err = false;
+		size_t skip = 0;
+		r_buf_seek (bf->buf, c->class_data_offset, R_BUF_SET);
+		dc->static_fields_size   = peek_uleb (bf->buf, &err, &skip);
+		dc->instance_fields_size = peek_uleb (bf->buf, &err, &skip);
+		dc->direct_methods_size  = peek_uleb (bf->buf, &err, &skip);
+		dc->virtual_methods_size = peek_uleb (bf->buf, &err, &skip);
+		if (err) {
+			free (dc);
+			goto beach;
 		}
-		c->class_data->static_fields_size = eof;
+		c->class_data = dc;
 
-		p = r_uleb128 (p, p_end - p, &eof);
-		if (p >= p_end) {
-			free (c->class_data);
-			return;
-		}
-		c->class_data->instance_fields_size = eof;
+		ut64 offset = c->class_data_offset + skip;
```

Comment:
why is this offset needed at all? it's probably enough to seek the rbuf to the right offset and there is no need to pass offset to other functions.